aws/resource_aws_s3_bucket_object.go

```diff
@@ -140,6 +140,13 @@ func resourceAwsS3BucketObject() *schema.Resource {
 				Type:         schema.TypeString,
 				Optional:     true,
 				ValidateFunc: validateArn,
+				DiffSuppressFunc: func(k, old, new string, d *schema.ResourceData) bool {
+					// ignore diffs where the user hasn't specified a kms_key_id but the bucket has a default KMS key configured
+					if new == "" && d.Get("server_side_encryption") == s3.ServerSideEncryptionAwsKms {
+						return true
+					}
+					return false
+				},
```

Comment:
To properly signal to Terraform that this schema attribute may be without a configuration value, but then have a value set in the state by the provider, we need to use `Computed: true` here. Future versions of Terraform core will be much more sensitive about returning an error when `Computed` is not used properly. 👍 

```go
Computed: true,
```

We should also add some documentation in `website/docs/r/s3_bucket_object.html.markdown` around this updated behavior:

```markdown
* `kms_key_id` - (Optional) Amazon Resource Name (ARN) of the KMS Key to use for object encryption. If the S3 Bucket has server-side encryption enabled, that value will automatically be used. If referencing the `aws_kms_key` resource, use the `arn` attribute. If referencing the `aws_kms_alias` data source or resource, use the `target_key_arn` attribute. Terraform will only perform drift detection if a configuration value is provided.
```